[REVIEW] Define and implement new rolling window APIs

harrism opened the pull request:
- [x] Define API for review
- [x] Implement for fixed-width types
- [ ] ~Implement UDFs for fixed-width types~ Defer to follow up PR
- [ ] ~Implement for strings (non-UDF versions)~  Defer to follow up PR.

Addresses part of #2945 

Merge after
 - [x] #3475
 - [x] #3487

trevorsm7 commented:
Jake mentioned that we can make `enum class` work, if you want to use that for rolling_operator

Don't forget to update the yml file with the new public headers

harrism commented:
enum class can work with Cython? @shwina do you know?

cwharris commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,92 @@
+                                       rolling_operator op,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  if (input.size() == 0)
```
out of curiosity: why avoid braces?

harrism replied:
I don't avoid braces, Mandrake, but I do deny them my essence.

jrhemstad commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,202 @@
+/** 
+ * @brief Rolling window aggregation operations
+ */
+enum rolling_operator {
```
```suggestion
enum class rolling_operator {
```

@shwina found a WAR for `enum class` in Cython.

jrhemstad commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,202 @@
+  SUM,       ///< Computes the sum of all values in the window
+  MIN,       ///< Computes minimum value in the window
+  MAX,       ///< Computes maximum value in the window
+  AVG,       ///< Computes arithmetic mean of all values in the window
```
We use `MEAN` in groupby (and I think other aggs). 
```suggestion
  MEAN,       ///< Computes arithmetic mean of all values in the window
```

harrism replied:
Not in rolling, apparently! (this was a copy-paste from the old API. But I agree MEAN is NICE, will change.

jrhemstad commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,202 @@
+ * - instead of storing NA/NaN for output rows that do not meet the minimum number of observations
+ *   this function updates the valid bitmask of the column to indicate which elements are valid.
+ *
+ * This function is asynchronous with respect to the GPU, i.e. the call will return before the
```
Do we want to have APIs like this?

jrhemstad replied:
I think this is probably true for other functions and it's just undocumented, which is probably a bad thing.

harrism replied:
@nsakharnykh had this comment in the original API. I don't think we want to document this as if it's a guarantee.

jrhemstad commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,202 @@
+ * @throws cudf::logic_error if window column type is not INT32
+ *
+ * @param[in] input_col The input column
+ * @param[in] window A column of INT32 window sizes. window[i] specifies window size for element i.
```
Non-nullable?

jrhemstad commented on cpp/tests/utilities/cudf_gtest.hpp:
```diff
@@ -92,3 +92,24 @@ struct TypeList<Types<TYPES...>> {
+
+// Utility for testing the expectation that an expression x throws the specified
+// exception whose what() message ends with the msg
+#define EXPECT_THROW_MESSAGE(x, exception, startswith, endswith)     \
```
It's good to test that an exception is thrown in exceptional cases, but I think testing for an exact error message is a bit too much on the [white-box](https://www.guru99.com/white-box-testing.html) end of the spectrum.

harrism replied:
You don't have to remember, you just have to run tests.

cwharris replied:
An alternative is declaring unique const error codes (integer) and expecting those. Could be static on the errors themselves,
```
cudf::specific_logic_error::error_id
```
or injectable
```
cudf::logic_error("message", cudf::error_codes::silly_mistake)
```

jrhemstad replied:
> You don't have to remember, you just have to run tests.

Is it really important to test the exact contents of the message?

harrism replied:
There's no other way to determine that an error was thrown for the reason you expect.

harrism replied:
Removed. :(

codereport commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+                 size_type min_periods)
+{
+  size_type i = blockIdx.x * block_size + threadIdx.x;
+  size_type stride = block_size * gridDim.x;
```
These should be initialised closer to where they are used aka below:
```cpp
    agg_op agg_operator{};
  gdf_size_type warp_valid_count{0};
```

harrism replied:
warp_valid_count must be initialized outside the loop. I figure the agg_op struct initialization will get optimized away, so I can move it.

codereport commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+  size_type i = blockIdx.x * block_size + threadIdx.x;
+  size_type stride = block_size * gridDim.x;
+
+    agg_op agg_operator{};
```
```suggestion
  agg_op agg_operator{};
```

codereport commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+                                                    rmm::mr::device_memory_resource *mr,
+                                                    cudaStream_t stream)
+  {
+    if (input.size() == 0) return empty_like(input);
```
Should `column_view` have an `is_empty` method? Obviously wouldn't be done in this PR

harrism replied:
Adding in this PR.

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+    // input data and mask
+  
+    std::vector<bitmask_type> in_valid; 
+    //std::tie(in_col, in_valid) = cudf::test::to_host<T>(input); 
```
```suggestion
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+  
+    std::vector<bitmask_type> in_valid; 
+    //std::tie(in_col, in_valid) = cudf::test::to_host<T>(input); 
+    in_valid = cudf::test::bitmask_to_host(input);
```
```suggestion
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+
+    // input data and mask
+  
+    std::vector<bitmask_type> in_valid; 
```
```suggestion
    auto in_valid = cudf::test::bitmask_to_host(input);
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+      for (size_type j = start_index; j < end_index; j++) {
+        if (!input.nullable() || cudf::bit_is_set(valid_mask, j))
+          count++;
+      }
```
Unclear to me why `!input.nullable()` is inside loop. If `j` does't impact `input` looks like the code should be something like:
```cpp
      size_type count = [&] () {
        if (!input.nullable())
          end_index - start_index;
        else
          for (size_type j = start_index; j < end_index; j++)
            if (cudf::bit_is_set(valid_mask, j))
              count++;
      } ();
```
Also, the `else` is pretty clearly a `count_if` but based on looking at `bitmask_type` being `uint8_t` and how `cudf::bit_is_set` works, I don't think we can use it.

rgsl888prabhu replied:
`bitmask_type` is `uint32_t`

https://github.com/rapidsai/cudf/blob/3d9da9e5da2e731d3d63a066a6d44fe3465cbb5a/cpp/include/cudf/types.hpp#L84

codereport replied:
sorry, yes that is what i meant

harrism replied:
I didn't want to diverge this code too far from the kernel that does the same thing on the GPU.  I like the first suggestion though, as it can work on both.

harrism replied:
BTW, we have exactly the function you want, called `cudf::count_set_bits(mask, start, end)`, but it expects a device pointer (runs a kernel). This is test code which runs on the CPU for verification.

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+          val = op(in_col[j], val);
+          count++;
+        }
+      }
```
Same comment as above.

harrism replied:
In this case there are two things in the loop -- the op() and the count. So moving the input.nullable outside the loop would cause me to have to have two loops.

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+  std::vector<TypeParam> col_data(num_rows);
+  std::vector<bool>      col_mask(num_rows);
+  std::fill(col_data.begin(), col_data.end(), 1);
+  std::fill(col_mask.begin(), col_mask.end(), 1);
```
You don't need to use `std::fill` here, you can use the _fill constructor_:
```cpp
  std::vector<TypeParam> col_data(num_rows, 1);
  std::vector<bool>      col_mask(num_rows, 1);
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+  size_type num_rows = 1000;
+
+  std::vector<TypeParam> col_data(num_rows);
+  std::vector<bool>      col_mask(num_rows);
```
```suggestion
  std::vector<bool>      col_mask(num_rows, 0);
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+
+  std::vector<TypeParam> col_data(num_rows);
+  std::vector<bool>      col_mask(num_rows);
+  std::fill(col_mask.begin(), col_mask.end(), 0);
```
```suggestion
```

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+  std::vector<TypeParam> col_data(num_rows);
+  std::vector<bool>      col_mask(num_rows);
+  std::fill(col_data.begin(), col_data.end(), 1);
+  std::fill(col_mask.begin(), col_mask.end(), 1);
```
Use _fill constructor_

codereport commented on cpp/tests/rolling/rolling_test.cpp:
```diff
@@ -0,0 +1,723 @@
+  std::vector<TypeParam> col_data(num_rows);
+  std::vector<bool>      col_mask(num_rows);
+  std::fill(col_data.begin(), col_data.end(), 1);
+  std::fill(col_mask.begin(), col_mask.end(), 1);
```
Use _fill constructor_

jrhemstad commented on cpp/src/rolling/rolling_detail.hpp:
```diff
@@ -0,0 +1,91 @@
+{
+  // return true if ColumnType is arithmetic type or
+  // AggOp is min_op/max_op/count_op for wrapper (non-arithmetic) types
+  template <typename ColumnType, class AggOp, bool is_mean>
```
The `is_mean` param seems redundant since `AggOp` is already available as compile time info.

harrism replied:
AggOp comes from device_operatores -- there's no mean operator. The division by count is computed on output (inherited from legacy implementation). So you can't tell that this is a mean from just the operator.

jrhemstad commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+    std::unique_ptr<column> output = (op == rolling_operator::COUNT) ?
+        make_numeric_column(cudf::data_type{cudf::INT32}, input.size(),
+                            cudf::UNINITIALIZED, stream, mr) :
+        allocate_like(input, cudf::experimental::mask_allocation_policy::ALWAYS, mr);
```
So the output type of a `MEAN` aggregation on an `int8_t` column will also be `int8_t`? I guess the user is expected to first cast the input to `float` if they want a `float` output? That sounds like it might be surprising and should at least be documented.

harrism replied:
I don't like it either. I think I will document it for this PR. We should add an output type in the future, possibly.

harrism replied:
Documented.

jrhemstad replied:
What groupby does is just fix the output type for some operations, like `MEAN` will always produce a `fp64` result, summing integers is always `int64`, etc.

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+    T val = agg_op::template identity<T>();
+    // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+    // for CUDA 10.0 and below (fixed in CUDA 10.1)
+    volatile cudf::size_type count = 0;
```
I assume, don't know how to  verify. @nsakharnykh wrote this. The kernel didn't change much.

jrhemstad replied:
If you remove the `volatile`, does it still produce correct results? It was giving wrong results w/o the `volatile` in the past.

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+  }
+
+  // sum the valid counts across the whole block  
+  gdf_size_type block_valid_count = 
```
Yikes, how did I miss this.

harrism commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,203 @@
+ * The returned column for `op == COUNT` always has INT32 type.
+ *
+ * @param[in] input_col The input column
+ * @param[in] window The static rolling window size.
```
Nope, pandas doesn't have a two-part  window. This naming comes from @nsakharnykh ...

nsakharnykh replied:
See above my suggestion to rename this to preceding and following window sizes, that would address both concerns about inclusiveness/exclusiveness and the direction. We would diverge from gQuant's (exclusive `window`, inclusive `forward_window`) and Panda's API (exclusive `window`, `centered` parameter), but will be more compatible with SQL.

harrism replied:
BTW, it was more than just renaming. I also had to change the behavior to compute the window as [i - preceding_window, i + following_window] instead of [i - window + 1, i + forward_window]. But it all works. Python will just have to change the bindings to account for this. @kkraus14 @shwina FYI.

harrism commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,203 @@
+  COUNT,     ///< Computes the number of values in the window
+  NUMBA_UDF, ///< A user-defined aggregation operation defined in PTX code generated by `numba`
+  CUDA_UDF,  ///< A generic aggregation operation defined in CUDA code
+  N_OPS,     ///< ALL NEW OPERATIONS SHOULD BE ADDED ABOVE THIS LINE
```
Nope.

harrism commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,203 @@
+ * notable differences:
+ * - instead of the center flag it uses the forward window size to allow for more flexible windows.
+ *   The total window size = window + forward_window. Element i uses elements
+ *   [i-window+1, i+forward_window] to do the window computation.
```
@nsakharnykh can you comment on this design choice? I'm afraid to change it lest we become incompatible with some driving use case.

nsakharnykh replied:
Pandas doesn't have `forward_window` parameter, instead they use `centered` to specify how the window is aligned with the current index. We went with `forward_window` because it give more flexibility (larger window to look back and smaller window to look forward?), and the decision to make `forward_window` inclusive / `window`. exclusive is based on gQuant's behavior (that we used as a starting point for the rolling windows design): https://github.com/rapidsai/gQuant/blob/f117268c3480a820ac8a93d90980192cec014883/gquant/cuindicator/rolling.py#L93. Looking at the SQL specs there are corresponding "preceding" and "following" parameters to control the window/forward_window sizes (https://www.red-gate.com/simple-talk/sql/learn-sql-server/window-functions-in-sql-server-part-2-the-frame/), and both are inclusive. I'm fine making the change to have both options to be inclusive, in fact, we could also rename the window sizes to `preceding_window_size`  and `following_window_size` to make it more clear.

harrism replied:
Thanks for the details! Will make these changes.

harrism commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,203 @@
+ * @param[in] window The static rolling window size.
+ * @param[in] forward_window The static window size in the forward direction.
+ * @param[in] min_periods Minimum number of observations in window required to have a value,
+ *                        otherwise 0 is stored in the valid bit mask for element i.
```
Done

harrism commented on cpp/include/cudf/rolling.hpp:
```diff
@@ -0,0 +1,203 @@
+ * @throws cudf::logic_error if window column type is not INT32
+ *
+ * @param[in] input_col The input column
+ * @param[in] window A non-nullable column of INT32 window sizes. window[i] specifies window size
```
Same response. Afraid to change this because I don't know the consequences.

harrism commented on cpp/src/rolling/rolling.cu:
```diff
@@ -0,0 +1,369 @@
+ * @tparam ColumnType  Datatype of values pointed to by the pointers
+ * @tparam agg_op  A functor that defines the aggregation operation
+ * @tparam is_mean Compute mean=sum/count across all valid elements in the window
+ * @tparam block_size CUDA block size for the kernel
```
Fixed.

harrism commented:
@jrhemstad I addressed all of your comments that I felt comfortable with. Could use feedback from @nsakharnykh on the original design for the `window` exclusivity and naming.

jrhemstad commented:
rerun tests

codereport commented on cpp/include/cudf/column/column_view.hpp:
```diff
@@ -104,6 +104,11 @@ class column_view_base {
+  /**---------------------------------------------------------------------------*
+   * @brief Returns true if `size()` returns zero, or false otherwise
+   *---------------------------------------------------------------------------**/
+  size_type empty() const noexcept { return size() == 0; }
```
I would highly recommend that we call this method `is_empty`. It is a widely acknowledged mistake that `empty` was the wrong name for the predicate method. It is unclear whether `empty` means `clear` (as in to empty the contents) or whether it is asking whether the data structure is empty (with either `true` or `false` response). All recent C++ features opt for the name `is_empty`, see:
https://en.cppreference.com/w/cpp/types/is_empty
https://en.cppreference.com/w/cpp/filesystem/is_empty

I even asked about this on twitter after listening to episode 73 of cppcast, and the recommendation was to always but `is_` or `has_` in front of predicates: https://twitter.com/hatcat01/status/1007161606994845697?s=20

harrism replied:
Thanks Conor, that's great info. Will change it.

harrism replied:
Done.

harrism commented:
rerun tests

jrhemstad commented:
rerun tests
